### PR TITLE
prov/rxd: initial overhaul of the rxd provider

### PR DESCRIPTION
**Pull request by aingerson**
This is a general restructuring of the rxd provider. Main changes include:

- Removing the excessive peer connection map and replacing it with a simple tx addr table and a rbmap to map the address to the core provider's fi_addr
- Flexible window size. Previously, the window size was 1, making exchanges very slow. The default is currently 128 messages but can be modified later to be changed via an environment variable or "smart" window, depending on other factors
- Currently, all data messages include a header to ensure it maps to the correct tx entry using the tx_id, rx_id, and key (assigned by the tx sequence). In the future, we are hoping to experiment with only including the header on retried messages. More experimentation needed with regards to this
- Disabling and removing tagged, RMA, and atomic capabilites

Passing benchmark latency/bandwidth and ubertest latency/bandwidth with complete data verification.

Still need to verify inject functions (passing benchmarks but not ubertest due to a lack of support for manual progress). But opening up for review anyway.

Signed-off-by: aingerson <alexia.ingerson@intel.com>

### CODE REVIEW

File: prov/rxd/src/rxd.h [hunk 1]
```diff
@@ -205,146 +165,93 @@ static inline struct rxd_cq *rxd_ep_rx_cq(struct rxd_ep *ep)
...
+			uint8_t source[];
+		};
 		struct {
-			uint64_t peer_msg_id;
-			uint8_t iov_count;
-			struct iovec src_iov[RXD_IOV_LIMIT];
-		} read_rsp;
+			struct rxd_data_hdr data_hdr;
+			uint8_t data[];
+		};
```
Comment (shefty) [inline prov/rxd/src/rxd.h]: There needs to be a common header for all packets.  That header should have a version field, plus any fields that are needed for all/most packets, including a type field to distinguish between data versus ctrl packets.  We need to ensure that protocol changes can be made without breaking communication with peers that may not have the same updates.  That is, you can't assume that the peer is running the same version of the library.

File: prov/rxd/src/rxd.h [hunk 2]
```diff
@@ -205,146 +165,93 @@ static inline struct rxd_cq *rxd_ep_rx_cq(struct rxd_ep *ep)
...
-
-	/* TODO: use iov and remove data copies */
-	char pkt_data[]; /* rxd_pkt_data*, followed by data */
+	fi_addr_t peer;
+	struct rxd_pkt pkt;
 };
 
+static inline int rxd_is_data_pkt(struct rxd_pkt_entry *pkt_entry)
+{
+	return (pkt_entry->pkt.resv != ~0);
```
Comment (shefty) [inline prov/rxd/src/rxd.h]: use a type field, not a reserved field that isn't actually reserved and must be set to a specific value.

File: prov/rxd/src/rxd.h [hunk 3]
```diff
@@ -205,146 +165,93 @@ static inline struct rxd_cq *rxd_ep_rx_cq(struct rxd_ep *ep)
...
-	struct dlist_entry entry;
-	struct rxd_tx_entry *tx_entry;
-	struct rxd_ep *ep;
 	struct fid_mr *mr;
-	int flags;
-
-	/* TODO: use iov and remove data copies */
-	char pkt_data[]; /* rxd_pkt_data*, followed by data */
+	fi_addr_t peer;
+	struct rxd_pkt pkt;
```
Comment (shefty) [inline prov/rxd/src/rxd.h]: pkt has a peer subfield.  Does that differ from the peer field immediately above?

File: prov/rxd/src/rxd.h [hunk 4]
```diff
@@ -205,146 +165,93 @@ static inline struct rxd_cq *rxd_ep_rx_cq(struct rxd_ep *ep)
...
 
 	struct dlist_entry entry;
 	struct dlist_entry pkt_list;
 
-	uint8_t op_type;
 	struct ofi_op_hdr op_hdr;
+	struct iovec iov[RXD_IOV_LIMIT];
+	uint8_t iov_count;
+};
+DECLARE_FREESTACK(struct rxd_x_entry, rxd_tx_fs);
```
Comment (shefty) [inline prov/rxd/src/rxd.h]: The rxm provider went to using multiple free pools for different packet types.  This allows each packet to be pre-formatted for its specific use case.  Consider following this model to reduce the per packet initialization overhead.

**Comment by jswaro** [comment]
> Disabling and removing tagged, RMA, and atomic capabilites

Out of curiosity, why are these being removed? I'd expect the utility providers to provide more functionality, not less.

**Comment by shefty** [comment]
The provider is being re-designed to improve stability and performance.  Those features will be re-added, but over the new design.

**Comment by jswaro** [comment]
> The provider is being re-designed to improve stability and performance. Those features will be re-added, but over the new design.

Makes sense. Thanks for clarifying.

File: prov/rxd/src/rxd_cq.c [hunk 5]
```diff
@@ -157,979 +158,394 @@ static int rxd_cq_write_tagged_signal(struct rxd_cq *cq,
...
-			credits = (rx_entry->msg_id == ctrl->msg_id) ?
-				  rx_entry->last_win_seg - rx_entry->exp_seg_no : 0;
-			rxd_ep_reply_ack(ep, ctrl, ofi_ctrl_ack, credits,
-				       ctrl->rx_key, peer->conn_data,
-				       ctrl->conn_id);
-			goto repost;
-		}
-	}
+	if (tx_entry->flags & RXD_NO_COMPLETION)
+		goto free;
```
Comment (shefty) [inline prov/rxd/src/rxd_cq.c]: nit: can drop goto, reverse if check, and move skipped code inside if

File: prov/rxd/src/rxd_cq.c [hunk 6]
```diff
@@ -157,979 +158,394 @@ static int rxd_cq_write_tagged_signal(struct rxd_cq *cq,
...
+			ret = rxd_handle_rts(ep, comp, pkt_entry);
+			break;
+		case RXD_CTS:
+			rxd_handle_cts(ep, comp, pkt_entry);
+			break;
+		case RXD_ACK:
+			rxd_handle_ack(ep, comp, pkt_entry);
+			break;
+		default:
+			assert(0);
```
Comment (shefty) [inline prov/rxd/src/rxd_cq.c]: don't trust peer data -- just discard with warning msg

File: prov/rxd/src/rxd_ep.c [hunk 7]
```diff
@@ -37,55 +37,94 @@
...
-	struct rxd_trecv_entry *trecv_entry;
+	struct dlist_entry *entry;
+	struct rxd_x_entry *rx_entry;
 	struct fi_cq_err_entry err_entry = {0};
 
 	ep = container_of(fid, struct rxd_ep, util_ep.ep_fid.fid);
-	fastlock_acquire(&ep->lock);
+	fastlock_acquire(&ep->util_ep.lock);
+
 	if (ep->util_ep.caps & FI_MSG) {
```
Comment (shefty) [inline prov/rxd/src/rxd_ep.c]: is this check needed?

**Pull request merged**